### PR TITLE
Restore update census mapbox cronjob

### PR DESCRIPTION
**Pull request by TurnerRiley**
Restores and updates the census mapbox cronjob following the work laid out in [this doc](https://docs.google.com/document/d/1ZcvnXAUfOwnmWdS8Z8JwwYDdHWTF6CrCWCmpgiKlw4U/edit#heading=h.rizng6chl91e):
 - Restores the [cronjob](https://github.com/code-dot-org/code-dot-org/blob/80777d646a9351de59404fbd173c67799c43dbda/bin/cron/update_census_mapbox)
 - Replaces hard-coded `censustiles` with `CDO.mapbox_upload_tileset`
 - Updates date logic in [bin/cron/update_census_mapbox#L45-L50](https://github.com/code-dot-org/code-dot-org/blob/80777d646a9351de59404fbd173c67799c43dbda/bin/cron/update_census_mapbox#L45-L50) to use a new `CURRENT_SCHOOL_YEAR` constant
 - Re-enables the cronjob by partially reverting [PR 50010](https://github.com/code-dot-org/code-dot-org/pull/50010)

## Links
Jira ticket: [here](https://codedotorg.atlassian.net/jira/software/c/projects/ACQ/boards/64?modal=detail&selectedIssue=ACQ-799&assignee=60d62161f65054006980bd71)
Access Report steps doc: [here](https://docs.google.com/document/d/1ZcvnXAUfOwnmWdS8Z8JwwYDdHWTF6CrCWCmpgiKlw4U/edit#heading=h.rizng6chl91e)

## Testing story
Because this cronjob has a dependency on the mapbox `tippecanoe` project, Suresh [suggested](https://codedotorg.slack.com/archives/C046G4TRLEN/p1696610379580919?thread_ts=1696608938.128709&cid=C046G4TRLEN) I set up an adhoc with a cloned production database. I did so and updated the adhoc's `locals.yml` file to include:
- `mapbox_upload_token`: [production's secret]
- `mapbox_access_token`: [production's secret]
- `mapbox_upload_tileset`: `censustiles-dev` (so that I wouldn't overwrite production's map tileset)
- `mapbox_access_tileset`: `censustiles-dev`

I then ran the cronjob on the adhoc (with the following output, including a printed `SUCCESS` indicating it reached the successful branch of the logic [here](https://github.com/code-dot-org/code-dot-org/pull/53972/commits/8d9f3f9848ee85aae2b3078a5ad2de821c6314a9#diff-5f35ebf46d1356726c591889ef47caf60877fce4f2326e6cf6331c81d3beb00fR62) (I included `puts` statements in this commit just to get more detailed output in case it failed)):
to_label

After this, I set my local `locals.yml` `mapbox_access_tileset` to `censustiles-dev` to view the updated map data locally and saw that the map had successfully updated. Looking at a higher level there are some pretty clear expected visual differences between before/after:

#### Previous
prev_some_no

#### New
new_more_no

And to check at a more granular level, I found a school whose `teaches_cs` field changed between last year and this year. Previously, Herman Avenue Elementary School was marked as having no data (white dot) but this year is marked as not supporting cs (purple dot):

#### Previous
prev_school_no_data

#### New
new_map_no

## PR Checklist:
- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: bin/cron/update_census_mapbox
Comment (TurnerRiley) [inline bin/cron/update_census_mapbox:12]: Is this needed?

File: bin/cron/update_census_mapbox
Comment (TurnerRiley) [inline bin/cron/update_census_mapbox]: Do we still want this since we're updating the different teaches_cs categories so there will probably be a lot of changes to this year?

File: bin/cron/update_census_mapbox
Comment (TurnerRiley) [inline bin/cron/update_census_mapbox:47]: Is this still true with using a constant school year? Do we want a constant school year or make it dynamic?

File: dashboard/test/models/census/census_summary_test.rb
Comment (TurnerRiley) [inline dashboard/test/models/census/census_summary_test.rb]: These tests (plus the similar ones on [lines 513-519](https://github.com/code-dot-org/code-dot-org/pull/53972/files#diff-a14752a56427bc9310e7734fb75aed35684a55eb87349095735ba8da0930d6ecR513-R519)) are new, and the rest are from the [original tests](https://github.com/code-dot-org/code-dot-org/commit/e47561190f834976b596665bab0ade4717191af8#diff-a14752a56427bc9310e7734fb75aed35684a55eb87349095735ba8da0930d6ec).

File: bin/cron/update_census_mapbox
Comment (davidsbailey) [reply bin/cron/update_census_mapbox]: good question! no, this check seems like a relic from when engineering owned the algorithm. the RED team will be responsible for ensuring the integrity of this data now.

**Comment by davidsbailey** [comment]
it looks like this PR is showing part of https://github.com/code-dot-org/code-dot-org/pull/53831 in the diff. can you please try merging staging into this branch again?

**Comment by TurnerRiley** [comment]
> it looks like this PR is showing part of #53831 in the diff. can you please try merging staging into this branch again?

Oh I didn't see any overlap between the two but I merged in staging just to be safe!

**Comment by davidsbailey** [comment]
> > it looks like this PR is showing part of #53831 in the diff. can you please try merging staging into this branch again?
> 
> Oh I didn't see any overlap between the two but I merged in staging just to be safe!

Thanks Turner, I'm glad you said something. I realized it only shows up when I try to view a subset of the commits (everything without the initial "restore cronjob" commit). Don't worry, I can work around it!

File: dashboard/app/models/census/census_summary.rb
Comment (davidsbailey) [inline dashboard/app/models/census/census_summary.rb]: apologies, it looks like I may have made an important omission in the work plan. the idea is that this cronjob should be uploading what's already in the table, exactly as provided by the seed_all task you recently added. what I'm seeing here is writing of new data to the census_summaries table, which looks like the old algorithm which we have gotten rid of and do not want to bring back. 

I may be missing something, but it looks like most or all of the code added to this file can be removed.

File: bin/cron/update_census_mapbox
Comment (davidsbailey) [inline bin/cron/update_census_mapbox]: it looks like this method call can be removed

**Comment by davidsbailey** [comment]
> Adding unit tests back in and running `./bin/cron/update_census_mapbox` without errors:

this seems like a good start, but I think we're going to want to do more thorough validation before merging this PR. I would recommend following all of the steps under "validate the data locally" in the doc. please be sure to double check that CDO.mapbox_upload_tileset returns censustiles-dev, so that you do not accidentally modify production data.

since you've potentially already uploaded data to the censustiles-dev , please see if you can confirm whether we are overwriting the tileset or just adding to it each time, possibly by finding the documentation on the mapbox site. if we are just adding to it, please see if you can find how to reset the data between uploads.

**Comment by TurnerRiley** [comment]
> > Adding unit tests back in and running `./bin/cron/update_census_mapbox` without errors:
> 
> this seems like a good start, but I think we're going to want to do more thorough validation before merging this PR. I would recommend following all of the steps under "validate the data locally" in the doc. please be sure to double check that CDO.mapbox_upload_tileset returns censustiles-dev, so that you do not accidentally modify production data.
> 
> since you've potentially already uploaded data to the censustiles-dev , please see if you can confirm whether we are overwriting the tileset or just adding to it each time, possibly by finding the documentation on the mapbox site. if we are just adding to it, please see if you can find how to reset the data between uploads.

In [this documentation](https://docs.mapbox.com/help/tutorials/upload-curl/), it says, "_The "tileset" value contains both your Mapbox username and the name that you want to call your data when it gets to Mapbox, like bike-trails, airports, or in this case traverse-bay. **If you use the name of an existing tileset, it will be overwritten**_" so I think we are overwriting the tileset!

**Comment by davidsbailey** [comment]
> In [this documentation](https://docs.mapbox.com/help/tutorials/upload-curl/), it says, "_The "tileset" value contains both your Mapbox username and the name that you want to call your data when it gets to Mapbox, like bike-trails, airports, or in this case traverse-bay. **If you use the name of an existing tileset, it will be overwritten**_" so I think we are overwriting the tileset!

great, nice find!

**Comment by sureshc** [comment]
I randomly noticed this Pull Request and vaguely recall that this job was causing significant performance issues on the production database because it was carrying out all of its work within a database transaction: https://github.com/code-dot-org/code-dot-org/blob/d0104bf2cf7437113680c19b2097b5469d827ae2/dashboard/app/models/census/census_summary.rb#L65-L93

@code-dot-org/infrastructure Will want to ensure that issue is addressed before this is merged. Carrying out a large batch of writes within a single transaction can negatively impact database performance and shouldn't be done unless all of the work needs to success or fail as a unit.

**Comment by davidsbailey** [comment]
Wow, great catch @sureshc ! FWIW, this cronjob is not what calls the seed_all method -- that is now done as part of the seed process, which was added in https://github.com/code-dot-org/code-dot-org/pull/53831 . Since that PR is already merged, it would probably make sense to open a separate PR to address that issue. Please let us know how we can help!

**Comment by davidsbailey** [comment]
> Carrying out a large batch of writes within a single transaction can negatively impact database performance and shouldn't be done unless all of the work needs to success or fail as a unit.

TLDR: I think we can simply remove the transaction wrapping the block.

long version
* this data is only updated once a year. for example, this fall we are adding 2023 data
* we don't display the new data until a bunch of manual verification is performed, and a code constant is updated to point to the new year
* based on these factors, there is relatively low risk of bad outcomes if the seed process fails and leaves partial data in the DB.

**Comment by TurnerRiley** [comment]
> > Carrying out a large batch of writes within a single transaction can negatively impact database performance and shouldn't be done unless all of the work needs to success or fail as a unit.
> 
> TLDR: I think we can simply remove the transaction wrapping the block.
> 
> long version
> 
> * this data is only updated once a year. for example, this fall we are adding 2023 data
> * we don't display the new data until a bunch of manual verification is performed, and a code constant is updated to point to the new year
> * based on these factors, there is relatively low risk of bad outcomes if the seed process fails and leaves partial data in the DB.

Thank you @sureshc and @davidsbailey for this feedback! I've removed the transaction wrapper and was able to successfully run the cronjob on an adhoc and view updated data! I'll be posting screenshots and a description shortly in the PR documentation.

**Comment by davidsbailey** [comment]
Nice work validating this change, Turner! per offline convo we are going to hold this for https://github.com/code-dot-org/code-dot-org/pull/54042, which will make sure that excluded schools in the 2023 data are in fact excluded from the map once we are showing 2023 data there.

File: dashboard/app/models/census/census_summary.rb
Comment (davidsbailey) [inline dashboard/app/models/census/census_summary.rb]: sorry but can you please split the changes to this file out into a separate PR? the cronjob and the seed change both have the potential to have unforeseen consequences, and it will be easiest to resolve if we have two separate PRs to work with so we can revert one easily.

Comment (TurnerRiley) [reply dashboard/app/models/census/census_summary.rb]: https://github.com/code-dot-org/code-dot-org/pull/54268

**Review by davidsbailey — APPROVED** [review]
Looks great Turner, thank you so much for your perseverance on this one!

Good to merge once this PR is merged:
* https://github.com/code-dot-org/code-dot-org/pull/54042

File: cookbooks/cdo-apps/templates/default/crontab.erb
Comment (TurnerRiley) [inline cookbooks/cdo-apps/templates/default/crontab.erb]: @davidsbailey Was just discussing this with Vijaya and I'm wondering if this cronjob frequency should be more infrequent than daily since this is theoretically something we only do once a year (plus perhaps some manual adjustments to values if the need arose)? Curious on your thoughts!

Comment (davidsbailey) [reply cookbooks/cdo-apps/templates/default/crontab.erb]: yes, great idea! I am wondering if it should actually be something we run manually and not be a cronjob at all? since you had to run it manually to validate the change, maybe the annual process would be to do a next step of manually updating the `censustiles` tileset which is visible in prod?

if we want to go that path, then the "cronjob" would move to the bin directory, and the census dev doc would be updated accordingly.

Comment (TurnerRiley) [reply cookbooks/cdo-apps/templates/default/crontab.erb]: Sounds good! I've moved the script to the bin directory and updated the dev doc to remove "cronjob" wording and say that the script must be run on production after the changes are merged. What do you mean by "next step of manually updating the `censustiles` tileset which is visible in prod"? Wouldn't we want to keep the production tileset as `censustiles` and just run the script on production once the changes are merged?

Comment (davidsbailey) [reply cookbooks/cdo-apps/templates/default/crontab.erb]: that sounds like a good way to do it! 

what I meant was, after validating locally, run the script locally again to update the `censustiles` tileset.

Comment (TurnerRiley) [reply cookbooks/cdo-apps/templates/default/crontab.erb]: Oh gotcha yes that makes sense! I don't think we can run this script locally because of `tippecanoe` (slack thread: [here](https://codedotorg.slack.com/archives/C046G4TRLEN/p1696608938128709)), so I think it would have to be run either in an adhoc or on production

File: bin/update_census_mapbox
Comment (davidsbailey) [inline bin/update_census_mapbox]: please remove these 2 lines, now that we are not running this as a cronjob

File: bin/update_census_mapbox
Comment (davidsbailey) [inline bin/update_census_mapbox]: please convert this warning to this check:
```suggestion
raise "CDO.mapbox_upload_token is not defined" unless CDO.mapbox_upload_token
```

 Based on your plan to run this in production, it shouldn't be necessary to follow the instructions in the production environment, which makes the comment a little confusing.

File: bin/update_census_mapbox
Comment (TurnerRiley) [reply bin/update_census_mapbox]: Perfect, I was just about to ask you if those lines were necessary!

**Review by davidsbailey — APPROVED** [review]
Looks good! Thanks for sticking with this one Turner!

**Pull request merged**
